### PR TITLE
easteregg/allow_non_authenticated_cr_clients_to_update

### PR DESCRIPTION
**Pull request by easteregg**

### CODE REVIEW

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 1]
```diff
@@ -98,11 +175,12 @@ const PersonalDetails = (() => {
     };
 
     const show_label_if_any_value   = ['account_opening_reason', 'citizen', 'place_of_birth', 'tax_residence', 'tax_identification_number'];
-    const force_update_fields       = ['tax_residence', 'tax_identification_number'];
+    const force_update_fields       = ['tax_residence', 'tax_identification_number'].concat(changeable_fields);
+    const has_changeable_fields     = Client.hasCostaricaAccount() && /^CR/.exec(Client.get('loginid')) && !is_fully_authenticated && !is_virtual;
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: this isn't global, but you're using it globally?

Comment (negarn) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: IMO we shouldn't be using `/^CR/.exec(Client.get('loginid'))`, we try not to hardcode login id as much as possible. just checking that current landing company shortcode is costarica can replace `Client.hasCostaricaAccount() && /^CR/.exec(Client.get('loginid'))`

Comment (negarn) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: then we can remove `!is_virtual` as well

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 2]
```diff
@@ -362,16 +471,16 @@ const PersonalDetails = (() => {
...
             init();
             const account_status = State.getResponse('get_account_status').status;
             get_settings_data = State.getResponse('get_settings');
             is_fully_authenticated = checkStatus(account_status , 'authenticated');
             if (is_virtual) {
                 getDetailsResponse(get_settings_data);
             }
-
-            if (!is_virtual || !residence) {
+            if (!is_virtual || !residence || has_changeable_fields) {
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: will there every be a case where client is virtual or doesn't have residence but has changeable fields? I think no, if client is virtual there are no changeable fields because we only allow CR to do that, and if client doesn't have residence the first thing they have to do is select residence before the rest of their details is displayed to them.

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 3]
```diff
@@ -98,11 +175,12 @@ const PersonalDetails = (() => {
...
 
     const show_label_if_any_value   = ['account_opening_reason', 'citizen', 'place_of_birth', 'tax_residence', 'tax_identification_number'];
-    const force_update_fields       = ['tax_residence', 'tax_identification_number'];
+    const force_update_fields       = ['tax_residence', 'tax_identification_number'].concat(changeable_fields);
+    const has_changeable_fields     = Client.hasCostaricaAccount() && /^CR/.exec(Client.get('loginid')) && !is_fully_authenticated && !is_virtual;
 
     const displayGetSettingsData = (get_settings, populate = true) => {
         Object.keys(get_settings).forEach((key) => {
-            const has_label         = show_label_if_any_value.includes(key);
+            const has_label         = show_label_if_any_value.includes(key) && !changeable_fields.includes(key);
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: please add a comment to explain this exception in the code

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 4]
```diff
@@ -143,6 +221,10 @@ const PersonalDetails = (() => {
                         if (force_update) { // Force pushing values, used for (API-)expected values
                             $element.attr({ 'data-force': true, 'data-value': el_value });
                         }
+                        // Update data-value on change for inputs
+                        if (should_update_value) {
+                            $(element_key).change(function () {this.setAttribute('data-value', this.value);});
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: why is this needed? doesn't existing behaviour handle change of value?

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 5]
```diff
@@ -199,6 +281,29 @@ const PersonalDetails = (() => {
                 { selector: '#chk_tax_id',     validations: is_financial ? [['req', { hide_asterisk: true, message: localize('Please confirm that all the information above is true and complete.') }]] : '', exclude_request: 1 },
             ];
 
+            // Push validations for changeable fields.
+            if (has_changeable_fields) {
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: I don't think we need an exception for pushing these to validations, we can include them in the validations list, since validator will handle if the selector for these doesn't exist or is invisible

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 6]
```diff
@@ -54,6 +58,76 @@ const PersonalDetails = (() => {
...
+            changeable_fields = changeable.only_before_auth;
+        }
+    };
+
+    /**
+     * Remove labels and static fields and replace them with input when fields are changeable.
+     *
+     * @param {get_settings} to prepopulate some of the values.
+     */
+    const displayChangeableFields = (get_settings) => {
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: can't we use `displayGetSettingsData` to also handle showing/hiding input for changeable fields? it feels like we're hardcoding a lot of fields in this function, while they are automated in `displayGetSettingsData`. it feels like repeating the work, but in a non-dynamic way.

File: src/templates/app/user/settings/personal_details.jsx [hunk 7]
```diff
@@ -36,8 +41,14 @@ const PersonalDetails = () => (
 
         <form className='form gr-padding-10 invisible' id='frmPersonalDetails'>
             <Fieldset legend={it.L('Details')}>
+                <Salutation className='invisible' />
+                <FirstName className='invisible' />
+                <LastName className='invisible' />
+                <DateOfBirth className='invisible' />
+                <Citizenship className='invisible' />
+                <AccountOpeningReason className='invisible' />
```
Comment (negarn) [inline src/templates/app/user/settings/personal_details.jsx]: usually we put the label next to the input row so we know that both exist, e.g. look at citizenship below

File: src/javascript/app/pages/user/account/settings/personal_details.js (hunk 1 above)
Comment (easteregg) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: Consider it done. Regarding global, I didn't get what you mean. Could you elaborate on that?

File: src/javascript/app/pages/user/account/settings/personal_details.js (hunk 4 above)
Comment (easteregg) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: Unfortunately no, `data-value` does not get updated when changing data (value does) but `form.handleSubmit` does not pick it up.

File: src/javascript/app/pages/user/account/settings/personal_details.js (hunk 5 above)
Comment (easteregg) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: Removing if statement. But I think letting the `forEach` loop stay, since it is more dynamic (with a few exceptions).

File: src/javascript/app/pages/user/account/settings/personal_details.js (hunk 2 above)
Comment (easteregg) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: I'll refactor this to separate the previous behavior from injecting `has_changeable_field`

File: src/templates/app/user/settings/personal_details.jsx (hunk 7 above)
Comment (easteregg) [reply src/templates/app/user/settings/personal_details.jsx]: done

File: src/javascript/app/pages/user/account/settings/personal_details.js (hunk 6 above)
Comment (easteregg) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: I can't think of a way that we can. Since I have to init data differently, we can't streamline it the way it has been done in `get_settings`. I would glad to refactor this if you have got some pointers to lead me to the right direction.

File: src/javascript/app/pages/user/account/settings/personal_details.js (hunk 4 above)
Comment (negarn) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: should we add that in a bigger scale rather than only fixing it in this file then?

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 8]
```diff
@@ -54,6 +59,70 @@ const PersonalDetails = (() => {
...
+        if (!has_changeable_fields) return;
+        changeable_fields.forEach(field => {
+            CommonFunctions.getElementById(`row_${field}`).setVisibility(1);
+        });
+
+        if (changeable_fields.includes('first_name') || changeable_fields.includes('last_name') || changeable_fields.includes('salutation')) {
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: would it be better to put these three in an array and then loop them? since adding a new field in the future would mean having to make it even longer

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 9]
```diff
@@ -54,6 +59,70 @@ const PersonalDetails = (() => {
...
+            CommonFunctions.getElementById('row_name').setVisibility(0);
+            CommonFunctions.getElementById('first_name').setVisibility(1);
+            CommonFunctions.getElementById('last_name').setVisibility(1);
+            CommonFunctions.getElementById('salutation').setVisibility(1);
+        }
+
+        if (changeable_fields.includes('date_of_birth')) {
+            const $input_el = $('#date_of_birth');
+            $input_el.setVisibility(1);
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: why not chain this with below?

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 10]
```diff
@@ -54,6 +59,70 @@ const PersonalDetails = (() => {
...
+                .change(function() {
+                    return CommonFunctions.dateValueChanged(this, 'date');
+                });
+
+            DatePicker.init({
+                selector: '#date_of_birth',
+                minDate : -100 * 365,
+                maxDate : (
+                    -18 * 365
+                ) - 5,
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: can you explain this?

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 11]
```diff
@@ -54,6 +59,70 @@ const PersonalDetails = (() => {
...
+                yearRange: '-100:-18',
+            });
+        }
+
+        const $options               = $('<div/>');
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: extra space before equal

Comment (negarn) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: should we put it inside the if? seems it's only used in there?

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 12]
```diff
@@ -113,7 +187,7 @@ const PersonalDetails = (() => {
                 if (populate) {
                     const should_update_value = /select|text/i.test(element_key.type);
                     if (has_label) {
-                        CommonFunctions.getElementById(`row_${element_id}`).setVisibility(1);
+                        CommonFunctions.getElementById(`row_lbl_${element_id}`).setVisibility(1);
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: how was it working before?

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 13]
```diff
@@ -362,38 +469,45 @@ const PersonalDetails = (() => {
...
             get_settings_data = State.getResponse('get_settings');
             is_fully_authenticated = checkStatus(account_status , 'authenticated');
+            has_changeable_fields = Client.get('landing_company_shortcode') === 'costarica' && !is_fully_authenticated;
+
             if (is_virtual) {
                 getDetailsResponse(get_settings_data);
+            } else if (has_changeable_fields) {
+                populateChangeableFields();
+                displayResidenceList();
+            } else if (!is_virtual || !residence) {
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: doesn't `!residence` take priority over `has_changeable_fields`?

Comment (negarn) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: we don't need to check `!is_virtual` here anymore, if client was virtual they would have gone to first if.

File: src/javascript/app/pages/user/account/settings/personal_details.js (hunk 12 above)
Comment (easteregg) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: template naming was not consistent. I've made a change to prefix all labels with `row_lbl`.

File: src/javascript/app/pages/user/account/settings/personal_details.js (hunk 13 above)
Comment (easteregg) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: I'll refactor to (is_virtual -> residence -> changeable_fields)

File: src/javascript/app/pages/user/account/settings/personal_details.js (hunk 10 above)
Comment (easteregg) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: I don't get it. Do you want me to add a comment in the code? or just explain it here?

Comment (negarn) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: comment in the code :)

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 14]
```diff
@@ -54,6 +59,70 @@ const PersonalDetails = (() => {
...
+        if (!has_changeable_fields) return;
+        changeable_fields.forEach(field => {
+            CommonFunctions.getElementById(`row_${field}`).setVisibility(1);
+        });
+
+        if (['first_name', 'last_name', 'salutation'].some(key => changeable_fields.includes(key))) {
+            CommonFunctions.getElementById('row_name').setVisibility(0);
+            CommonFunctions.getElementById('first_name').setVisibility(1);
+            CommonFunctions.getElementById('last_name').setVisibility(1);
+            CommonFunctions.getElementById('salutation').setVisibility(1);
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: I meant if we loop the array, we can combine these three lines to `CommonFunctions.getElementById(looping_item).setVisibility(1);`, and handle setting row_name to invisible once

Comment (negarn) [reply src/javascript/app/pages/user/account/settings/personal_details.js]: but if you think this is better then keep it

File: src/javascript/app/pages/user/account/settings/personal_details.js [hunk 15]
```diff
@@ -21,11 +23,14 @@ const PersonalDetails = (() => {
...
         residence,
-        get_settings_data;
+        get_settings_data,
+        has_changeable_fields,
+        changeable_fields;
 
     const init = () => {
         editable_fields   = {};
         get_settings_data = {};
+        changeable_fields = [];
```
Comment (negarn) [inline src/javascript/app/pages/user/account/settings/personal_details.js]: shouldn't we add tax res and tax id here until API returns it?

File: src/sass/app/websocket_pages.scss [hunk 16]
```diff
@@ -283,3 +283,9 @@
         justify-content: center;
     }
 }
+
+.realws-content {
+    div.select-dropdown {
+        width: 100%;
```
Comment (negarn) [inline src/sass/app/websocket_pages.scss]: we have some very long state names, or country names, I can't remember which. sometimes 100% causes it to become too long. please check this,

Comment (easteregg) [reply src/sass/app/websocket_pages.scss]: Yeah, countries, this doesn't affect that one. Thanks for pointing that out. I'll double check the mobile ones.
